amf-api-contract/shared/src/main/scala/amf/apicontract/internal/validation/shacl/graphql/GraphQLValidator.scala

```diff
@@ -1,18 +1,60 @@
 package amf.apicontract.internal.validation.shacl.graphql
 
 import amf.core.client.scala.model.DataType
+import amf.core.client.scala.model.domain.extensions.PropertyShape
 import amf.core.client.scala.model.domain.{DataNode, NamedDomainElement, ScalarNode, Shape}
 import amf.core.internal.metamodel.Field
 import amf.core.internal.metamodel.domain.ScalarNodeModel
-import amf.core.internal.metamodel.domain.extensions.DomainExtensionModel
+import amf.core.internal.metamodel.domain.extensions.{DomainExtensionModel, PropertyShapePathModel}
 import amf.core.internal.parser.domain.Annotations
 import amf.shapes.client.scala.model.domain._
+import amf.shapes.client.scala.model.domain.federation.Key
 import amf.shapes.client.scala.model.domain.operations.AbstractParameter
 import amf.shapes.internal.domain.metamodel.NodeShapeModel
 import amf.shapes.internal.domain.metamodel.operations.AbstractParameterModel
 import amf.validation.internal.shacl.custom.CustomShaclValidator.ValidationInfo
 
 object GraphQLValidator {
+
+  def checkValidPath(path: Seq[PropertyShape], key: Key): Seq[ValidationInfo] = {
+    path.flatMap { propertyShape =>
+      propertyShape.range match {
+        case n: NodeShape =>
+          if (n.isAbstract.value())
+            validationInfo(
+              PropertyShapePathModel.Path,
+              s"Property '${propertyShape.name}' reference by field set can't be from an interface",
+              key.annotations
+            )
+          else if (n.isInputOnly.value())
+            validationInfo(
+              PropertyShapePathModel.Path,
+              s"Property '${propertyShape.name}' reference by field set can't be from an input type",
+              key.annotations
+            )
+          else {
+            if (path.size == 1) {
+              validationInfo(
+                PropertyShapePathModel.Path,
+                s"Property '${propertyShape.name}' reference by field set can't be an object type",
+                key.annotations
+              )
+            } else None
+          }
+        case _ => None
```

Comment:
You need to throw a validation when matching other shapes (eg. UnionShape) except ScalarShape. If I remember correctly all intermediate steps of a PropertyShapePath will be validated as NodeShapes during linking. All except the last propertyShapeRange. That must be a ScalarShape if I'm not mistaken